### DISCUSSION TITLE
MILESTONE RELEASE: v0.13.1 (PREVIEW 2) -- 64 BIT SUPPORT!

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#496] [General]
### As per the title, v0.13.1 now includes full compilation and debugging support for twinBASIC projects.

To get started, first you need to create a win64 build configuration in your projects.  To help you, when you first open your projects in v0.13.1, the IDE will prompt you to create one:

<img width="340" alt="tb_win64Notice" src="https://user-images.githubusercontent.com/25738109/142645665-56281a5e-334f-41da-86eb-c0533212d943.png">

You should now see the build configuration appear in the explorer:
<img width="289" alt="tb_win64Config" src="https://user-images.githubusercontent.com/25738109/142645712-8592835c-eb27-4dd2-af96-98e02fe3f9e7.png">

Now you should be able to activate the win64 build configuration:
<img width="265" alt="tb_changeActiveBuild" src="https://user-images.githubusercontent.com/25738109/142645872-56eac113-e06f-425d-9854-bf192dc0dccf.png">

After selecting win64, you'll be prompted to restart the compiler.  After doing so, you'll be debugging and compiling in full glorious 64-bit.  Have a nice weekend folks!

**Comment by WaynePhillipsEA** [comment]
v0.13.2 fixes the 'Active Build' drop down list to be automatically refreshed as build configurations get updated / modified in the explorer, without requiring a restart of VS code.

**Comment by mansellan** [comment]
It's an interesting quirk of history that there was never a 64-bit Classic VB. By 2010, Microsoft had a compiler and debugger for it, for Office VBA7. All they needed to do was add a linker, and they could have released VB7. I guess at that point, they wanted classic VB to die and everyone to move to VB.Net. I wonder if MS engineers argued for it and management / sales overruled them. Pure speculation on my part.

In hindsight, perhaps its a good thing, it inspired Wayne to create twinBASIC with all the modernity it's bringing :-)

EDIT: I'm wrong, they weren't as close as that - see below.

**Reply by WaynePhillipsEA** [reply]
I think there was more work required than just the linker, as VBA doesn't compile to native code.  I speculate that re-writing/upgrading the old native code compiler would have been a big job on its own.

**Reply by mansellan** [reply]
I thought VBA got compiled into 2 stages, both stored in the persisited file. Firstly into p-code, which is actually what gets decompiled into the VBE source code window. But latterly into actual native code, which only works if the machine bitness and Office version match? And if that fails, it just recompiles stage 2 on the target machine.

I'm sure I'm wrong, but I don't know why?

**Reply by WaynePhillipsEA** [reply]
VBA code gets stored in three versions.  
1) the raw source code (compressed) 

2) the tokenized version of the source code that the IDE uses for live editing.  This is practically a 1:1 version of the raw source code, but in a tokenized form.

3) the 'compiled' version that Microsoft internally calls ExCode.  (This is the version that is stored in compiled ACCDE and MDE files, with the other two forms stripped out.)

Important to note is that the third form is NOT proper native-code.  It is a form that runs through a virtual machine.  In VB6 terms, the third form is equivalent to compiling an EXE with the 'pcode' option of the VB6 compiler.   VBA never gets turned into fully compiled nativecode - it always runs as ExCode (or 'pcode' in VB6 terms).

**Reply by mansellan** [reply]
Many thanks for the explanation, I knew I had it a little wrong.

**Reply by mansellan** [reply]
Just curious @WaynePhillipsEA - are you aware of the old Developer editions of Office (97 and XP)? They could produce a subset of the output types available in VB6. IIRC, DLLs and ActiveX. I'm guessing that buried somewhere deep inside the VBA libraries is some compiler code from VB6 (and likely only 32-bit aware).
